### DISCUSSION TITLE
source-changes explained

### DISCUSSION THREAD

**Discussion by usox** [ampache/ampache#2689] [Show and tell]
### Namespace structure

Q'n'd the current namespace structure in the source-changes branch for a better understanding. These namespaces are not final, though.
The main goal is to separate the actions which come from different entry points (a click in the gui, an api method call, ...) from the logic and the database layer. The separation of model and db-access will make the potential transition to an orm in future much easier.

- Ampache\Config: Contains config and initialization related classes (application bootstrapping, dic creation, ...)
- Ampache\Gui: Contains classes related to the new template system for the `legacy` gui
- Ampache\Model: Contains all model classes (like Song, Artist, Art, ...)
- Ampache\Module: Contains the application logic separated in single domains (e.g. "Ampache\Module\Album" will contain the application logic for all album related actions)
- Ampache\Module\Api: Api related classes (api methods, ...)
- Ampache\Module\Application: Separate classes for every action which can be performed in the "legacy" gui (deleting an album, ...)
- Ampache\Plugin: Contains all installable plugins
- Ampache\Repository: Contains classes which perform direct database access (crud actions, selects...which are currently directly performed within the Model-classes)

### Repositories, models and application logic

Currently, most of the application logic takes place within the model files (e.g. User::disable does not just set the disable state for the user, it also deletes existing sessions). As the model classes should whether contain logic nor directly access the database - it has to be split up. Another reason for the split is the testability of the code and the preparation for a possible shift to a real `orm` in future.

- Repositories: A repository contains all methods which directly access the database to retrieve/manipulate data. They should not contain application logic and should return whether raw data or the corresponding models. Every table in the database has its own repository
-  Models: Used to represent the data of a certain database table. If used as `active records`, they're some kind of `transport media` to transport the data between the application logic layer and the repositories (https://en.wikipedia.org/wiki/Active_record_pattern)
- Application Logic: The logic parts of the old models got/will get moved into separate classes. This encapsulation makes the logic testable, reduces dependencies and reduces the overall class size (e.g. in develop, the User class has about 1800 lines) which makes it better maintainable.

**Comment by AshotN** [comment]
Is 'Legacy' referring to the Reborn interface we currently have?

**Reply by usox** [reply]
> Is 'Legacy' referring to the Reborn interface we currently have?

Yepp. Maybe `legacy` sounds too negative, though.

**Reply by lachlan-00** [reply]
In most things it's referred to as web ui.

**Reply by AshotN** [reply]
> > Is 'Legacy' referring to the Reborn interface we currently have?
> 
> Yepp. Maybe `legacy` sounds too negative, though.

The legacy name works, just wanted to clarify.

**Comment by usox** [comment]
Added a short note about repositories and model classes
